images/virtualization-artifact/pkg/controller/vmop/internal/lifecycle.go

```diff
@@ -0,0 +1,192 @@
+	if changed.Status.Phase == "" {
+		// TODO add label with vm name.
+		changed.Status.Phase = virtv2.VMOPPhasePending
+		// Add all conditions in unknown state.
+		for _, condType := range lifeCycleConditions {
+			service.SetCondition(metav1.Condition{
+				Type:   condType,
+				Status: metav1.ConditionUnknown,
+			}, &changed.Status.Conditions)
+		}
+	}
+
+	// Ignore if VMOP is in final state.
+	if IsFinalState(changed) {
+		return reconcile.Result{}, nil
+	}
+
+	// Get VM for Pending and InProgress checks.
+	vm, err := s.VirtualMachine(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("get VirtualMachine for VMOP: %w", err)
+	}
+	if vm == nil {
+		changed.Status.Phase = virtv2.VMOPPhaseFailed
+		service.SetCondition(metav1.Condition{
+			Type:    vmopcondition.CompletedType,
+			Reason:  vmopcondition.VirtualMachineNotFound,
+			Status:  metav1.ConditionFalse,
+			Message: "VirtualMachine not found",
+		}, &changed.Status.Conditions)
+		return reconcile.Result{}, nil
+	}
+
+	if changed.Status.Phase == virtv2.VMOPPhaseInProgress {
+		h.logger.Debug("Operation in progress, check if VM is completed", "vm.phase", vm.Status.Phase, "vmop.phase", changed.Status.Phase)
+		return h.checkOperationComplete(changed, vm)
+	}
+
+	// At this point VMOP is in Pending phase, do some validation checks.
+
+	// Fail if there is at least one other VirtualMachineOperation in progress.
+	found, err := h.vmopSrv.OtherVMOPIsInProgress(ctx, changed)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if found {
+		changed.Status.Phase = virtv2.VMOPPhaseFailed
+		service.SetCondition(metav1.Condition{
+			Type:    vmopcondition.CompletedType,
+			Reason:  vmopcondition.OtherOperationsAreInProgress,
+			Status:  metav1.ConditionFalse,
+			Message: "Other VirtualMachineOperations are in progress",
+		}, &changed.Status.Conditions)
+		return reconcile.Result{}, nil
+	}
+
+	// Fail if VirtualMachineOperation is not applicable for run policy.
+	if !h.vmopSrv.IsApplicableForRunPolicy(changed, vm.Spec.RunPolicy) {
+		changed.Status.Phase = virtv2.VMOPPhaseFailed
+		service.SetCondition(metav1.Condition{
+			Type:    vmopcondition.CompletedType,
+			Reason:  vmopcondition.NotApplicableForRunPolicy,
+			Status:  metav1.ConditionFalse,
+			Message: fmt.Sprintf("Operation type %s is not applicable for runPolicy %s", changed.Spec.Type, vm.Spec.RunPolicy),
+		}, &changed.Status.Conditions)
+		return reconcile.Result{}, nil
+	}
+
+	// Fail if VirtualMachineOperation is not applicable for VM phase.
+	if !h.vmopSrv.IsApplicableForVMPhase(changed, vm.Status.Phase) {
+		changed.Status.Phase = virtv2.VMOPPhaseFailed
+		service.SetCondition(metav1.Condition{
+			Type:    vmopcondition.CompletedType,
+			Reason:  vmopcondition.NotApplicableForVMPhase,
+			Status:  metav1.ConditionFalse,
+			Message: fmt.Sprintf("Operation type %s is not applicable for vm phase %s", changed.Spec.Type, vm.Status.Phase),
```

Comment:
Avoid using abbreviations for user messages

Reply:
Rewritten

			Message: fmt.Sprintf("Operation type %s is not applicable for VirtualMachine with runPolicy %s", changed.Spec.Type, vm.Spec.RunPolicy),
			Message: fmt.Sprintf("Operation type %s is not applicable for VirtualMachine in phase %s", changed.Spec.Type, vm.Status.Phase),